---
title: "scRNAseq analysis"
author: "Sergi Roig"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: show
    toc: true
    toc_depth: 3
    theme: cosmo
    highlight: textmate
    number_sections: true
  pdf_document:
    latex_engine: xelatex
    toc: true
    toc_depth: '3'
  word_document:
    toc: true
    toc_depth: '3'
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
## Global options
library(knitr)
library(rmdformats)
library(tinytex)
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=75)
knitr::opts_chunk$set(echo = TRUE, tidy = TRUE, tidy.opts = list(width.cutoff = 60))
```

```{r packages, message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(SingleR)
library(dplyr)
library(celldex)
library(RColorBrewer)
library(SingleCellExperiment)
library(patchwork)
library(reshape2)
library(hdf5r)
```

```{r functions, message=FALSE, warning=FALSE, include=FALSE}
# function plot_integrated_clusters:
plot_integrated_clusters = function (srat) { 
  ## take an integrated Seurat object, plot distributions over orig.ident
  count_table <- table(srat@meta.data$seurat_clusters, srat@meta.data$orig.ident)
  count_mtx   <- as.data.frame.matrix(count_table)
  count_mtx$cluster <- rownames(count_mtx)
  melt_mtx    <- melt(count_mtx)
  melt_mtx$cluster <- as.factor(melt_mtx$cluster)
  
  cluster_size   <- aggregate(value ~ cluster, data = melt_mtx, FUN = sum)
  
  sorted_labels <- paste(sort(as.integer(levels(cluster_size$cluster)),decreasing = T))
  cluster_size$cluster <- factor(cluster_size$cluster,levels = sorted_labels)
  melt_mtx$cluster <- factor(melt_mtx$cluster,levels = sorted_labels)
  
  colnames(melt_mtx)[2] <- "dataset"
  
  
  p1 <- ggplot(cluster_size, aes(y= cluster,x = value)) + geom_bar(position="dodge", stat="identity",fill = "grey60") + 
    theme_bw() + scale_x_log10() + xlab("Cells per cluster, log10 scale") + ylab("")
  p2 <- ggplot(melt_mtx,aes(x=cluster,y=value,fill=dataset)) + 
    geom_bar(position="fill", stat="identity") + theme_bw() + coord_flip() + 
    scale_fill_brewer(palette = "Set2") +
    ylab("Fraction of cells in each dataset") + xlab("Cluster number") + theme(legend.position="top")
  
  p2 + p1 + plot_layout(widths = c(3,1))
  
  
}


plot_integrated_clusters_annotated <- function(srat) { 
  # Tabla: cluster_annotation x orig.ident
  count_table <- table(srat@meta.data$cluster_annotation, srat@meta.data$orig.ident)
  count_mtx   <- as.data.frame.matrix(count_table)
  count_mtx$cluster <- rownames(count_mtx)
  melt_mtx    <- melt(count_mtx)
  melt_mtx$cluster <- as.factor(melt_mtx$cluster)
  
  # Total por anotación
  cluster_size <- aggregate(value ~ cluster, data = melt_mtx, FUN = sum)
  
  # Ordenar por tamaño de clúster
  sorted_labels <- cluster_size$cluster[order(cluster_size$value, decreasing = TRUE)]
  cluster_size$cluster <- factor(cluster_size$cluster, levels = sorted_labels)
  melt_mtx$cluster     <- factor(melt_mtx$cluster, levels = sorted_labels)
  
  colnames(melt_mtx)[2] <- "dataset"
  
  # Gráfico 1: tamaño absoluto
  p1 <- ggplot(cluster_size, aes(y = cluster, x = value)) + 
    geom_bar(position = "dodge", stat = "identity", fill = "grey60") + 
    theme_bw() + 
    scale_x_log10() + 
    xlab("Cells per cluster (log10)") + 
    ylab("")
  
  # Gráfico 2: proporciones por `orig.ident`
  p2 <- ggplot(melt_mtx, aes(x = cluster, y = value, fill = dataset)) + 
    geom_bar(position = "fill", stat = "identity") + 
    theme_bw() + 
    coord_flip() + 
    scale_fill_brewer(palette = "Set2") +
    ylab("Fraction of cells per dataset") + 
    xlab("Annotated cluster") + 
    theme(legend.position = "top")
  
  # Combina ambos plots
  p2 + p1 + plot_layout(widths = c(3, 1))
}

```



# Data loading

```{r data loading}
# read the filtered Cell Ranger h5 matrices
spl_WT <- Read10X_h5("DATA/WT_SPsample_filtered_feature_bc_matrix.h5")
spl_KO <- Read10X_h5("DATA/KO_SP_sample_filtered_feature_bc_matrix.h5")
bm_WT  <- Read10X_h5("DATA/WT_BM_sample_filtered_feature_bc_matrix.h5")
bm_KO  <- Read10X_h5("DATA/KO_BM_sample_filtered_feature_bc_matrix.h5")

# create the appropriate Seurat
spl_WT_obj <- CreateSeuratObject(spl_WT, project = "spl_WT")
spl_KO_obj <- CreateSeuratObject(spl_KO, project = "spl_KO")
bm_WT_obj  <- CreateSeuratObject(bm_WT, project = "bm_WT")
bm_KO_obj  <- CreateSeuratObject(bm_KO, project = "bm_KO")

# let’s remove matrices to save memory
rm(spl_WT)
rm(spl_KO)
rm(bm_WT)
rm(bm_KO)
```

# Preliminar filter

## WT Spleen

```{r filtering plots WT spleen}
spl_WT_obj[["percent.mt"]]  <- PercentageFeatureSet(spl_WT_obj, pattern = "^mt-")
spl_WT_obj[["percent.rbp"]] <- PercentageFeatureSet(spl_WT_obj, pattern = "^Rps|^Rpl")
VlnPlot(spl_WT_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rbp"), ncol = 4)
```

## PrimPol KO spleen

```{r filtering plots KO spleen}
spl_KO_obj[["percent.mt"]]  <- PercentageFeatureSet(spl_KO_obj, pattern = "^mt-")
spl_KO_obj[["percent.rbp"]] <- PercentageFeatureSet(spl_KO_obj, pattern = "^Rps|^Rpl")
VlnPlot(spl_KO_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rbp"), ncol = 4)
```

## WT bone marrow

```{r filtering plots WT BM}
bm_WT_obj[["percent.mt"]]  <- PercentageFeatureSet(bm_WT_obj, pattern = "^mt-")
bm_WT_obj[["percent.rbp"]] <- PercentageFeatureSet(bm_WT_obj, pattern = "^Rps|^Rpl")
VlnPlot(bm_WT_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rbp"), ncol = 4)
```

## PrimPol KO bone marrow

```{r filtering plots KO BM}
bm_KO_obj[["percent.mt"]]  <- PercentageFeatureSet(bm_KO_obj, pattern = "^mt-")
bm_KO_obj[["percent.rbp"]] <- PercentageFeatureSet(bm_KO_obj, pattern = "^Rps|^Rpl")
VlnPlot(bm_KO_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rbp"), ncol = 4)
```

# Quick filtering

```{r filtering}
# quick filtering
spl_WT_obj <- subset(spl_WT_obj, subset = nFeature_RNA > 250 & nFeature_RNA < 5000 & percent.mt < 20)
spl_KO_obj <- subset(spl_KO_obj, subset = nFeature_RNA > 500 & nFeature_RNA < 5000 & percent.mt < 15)
bm_WT_obj <- subset(bm_WT_obj, subset = nFeature_RNA > 250 & nFeature_RNA < 5000 & percent.mt < 25)
bm_KO_obj <- subset(bm_KO_obj, subset = nFeature_RNA > 500 & nFeature_RNA < 5000 & percent.mt < 25)
```

# Spleen

## Integration, normalization

```{r}
# creat a list
spl_list <- list()
spl_list[["spl_WT"]] <- spl_WT_obj
spl_list[["spl_KO"]] <- spl_KO_obj

# normalize/find HVG
for (i in 1:length(spl_list)) {
  spl_list[[i]] <- NormalizeData(spl_list[[i]], verbose = F)
  spl_list[[i]] <- FindVariableFeatures(spl_list[[i]], selection.method = "vst", nfeatures = 2000, verbose = F)
}
# to find integration anchors and actually perform integration
# it takes time...
spl_anchors <- FindIntegrationAnchors(object.list = spl_list, dims = 1:30)
spl_seurat <- IntegrateData(anchorset = spl_anchors, dims = 1:30)
# to save ram
rm(spl_list)
rm(spl_anchors)

# before integration
DefaultAssay(spl_seurat) <- "RNA"
spl_seurat <- NormalizeData(spl_seurat, verbose = F)
spl_seurat <- FindVariableFeatures(spl_seurat, selection.method = "vst", nfeatures = 2000, verbose = F)
spl_seurat <- ScaleData(spl_seurat, verbose = F)
spl_seurat <- RunPCA(spl_seurat, npcs = 30, verbose = F)
spl_seurat <- RunUMAP(spl_seurat, reduction = "pca", dims = 1:30, verbose = F)

DimPlot(spl_seurat,reduction = "umap") +
  plot_annotation(title = "Spleen, before integration")
```

Integrated:

```{r}
# the integrated assay (it’s already normalized and HVGs are selected)
DefaultAssay(spl_seurat) <- "integrated"
spl_seurat <- ScaleData(spl_seurat, verbose = F)
spl_seurat <- RunPCA(spl_seurat, npcs = 30, verbose = F)
spl_seurat <- RunUMAP(spl_seurat, reduction = "pca", dims = 1:30, verbose = F)

DimPlot(spl_seurat, reduction = "umap") + 
  plot_annotation(title = "Spleen, after integration (Seurat 3)")
```

If we split:

```{r}
# split
DimPlot(spl_seurat, reduction = "umap", split.by = "orig.ident") + NoLegend()
```

## Clusterization

```{r}
# let's cluster
spl_seurat <- FindNeighbors(spl_seurat, dims = 1:30, k.param = 10, verbose = F)
spl_seurat <- FindClusters(spl_seurat, verbose = F, resolution = 0.1)
DimPlot(spl_seurat,label = T) + NoLegend()
```

## Annotation


### Immuno dataset

```{r}
# Extraer las matrices de cada muestra
expr_WT <- GetAssayData(spl_seurat, assay = "RNA", layer = "data.spl_WT")
expr_KO <- GetAssayData(spl_seurat, assay = "RNA", layer = "data.spl_KO")

# Combina columnas (células) de ambos
expr <- cbind(expr_WT, expr_KO)


spl_sce <- SingleCellExperiment(
  assays = list(logcounts = expr),
  colData = spl_seurat@meta.data
)


# mouse inmune reference
ref <- celldex::ImmGenData()

# Corre SingleR para anotar
pred_spl <- SingleR(test = spl_sce, ref = ref, labels = ref$label.fine)

# Añade la anotación a tu objeto Seurat
spl_seurat$SingleR_labels <- pred_spl$labels

# Acorta los labels (esto es solo un ejemplo, puedes adaptarlo)
spl_seurat$SingleR_clean <- gsub("\\s*\\([^\\)]+\\)", "", spl_seurat$SingleR_labels)

# Visualiza con nombres más cortos
DimPlot(spl_seurat, group.by = "SingleR_clean", label = TRUE, repel = TRUE) + NoLegend()
```

Para anotar los clusteres:

```{r}
## para anotar los clusteres

# 1. Extraer metadata
meta <- spl_seurat@meta.data

# 2. Crear tabla con la anotación más frecuente por clúster
annot_df <- meta %>%
  mutate(cluster = as.character(seurat_clusters)) %>%
  group_by(cluster) %>%
  count(SingleR_clean) %>%
  slice_max(n, n = 1, with_ties = FALSE) %>%
  ungroup()

# 3. Crear un vector de anotación por clúster
annot_vec <- setNames(annot_df$SingleR_clean, annot_df$cluster)

# 4. Asignar anotaciones a cada célula usando su clúster
meta$cluster_annotation <- annot_vec[as.character(meta$seurat_clusters)]

# 5. Asegurarse de que se guarda en el objeto Seurat
spl_seurat@meta.data <- meta

DimPlot(spl_seurat, group.by = "cluster_annotation", label = TRUE, repel = TRUE) + NoLegend()
```


```{r}
# ploteamos los porcentajes
plot_integrated_clusters_annotated(spl_seurat)
```

### Mousernaseq dataset

Podemos usar otro tipo de referencia, como MouseRNASeqData:

```{r}
# mouse inmune reference
ref <- celldex::MouseRNAseqData()

# Corre SingleR para anotar
pred_spl <- SingleR(test = spl_sce, ref = ref, labels = ref$label.fine)

# Añade la anotación a tu objeto Seurat
spl_seurat$SingleR_labels <- pred_spl$labels

# Acorta los labels (esto es solo un ejemplo, puedes adaptarlo)
spl_seurat$SingleR_clean <- gsub("\\s*\\([^\\)]+\\)", "", spl_seurat$SingleR_labels)

# Visualiza con nombres más cortos
DimPlot(spl_seurat, group.by = "SingleR_clean", label = TRUE, repel = TRUE) + NoLegend()
```

Anotamos los clusteres:

```{r}
# para anotar los clusteres

# 1. Extraer metadata
meta <- spl_seurat@meta.data

# 2. Crear tabla con la anotación más frecuente por clúster
annot_df <- meta %>%
  mutate(cluster = as.character(seurat_clusters)) %>%
  group_by(cluster) %>%
  count(SingleR_clean) %>%
  slice_max(n, n = 1, with_ties = FALSE) %>%
  ungroup()

# 3. Crear un vector de anotación por clúster
annot_vec <- setNames(annot_df$SingleR_clean, annot_df$cluster)

# 4. Asignar anotaciones a cada célula usando su clúster
meta$cluster_annotation <- annot_vec[as.character(meta$seurat_clusters)]

# 5. Asegurarse de que se guarda en el objeto Seurat
spl_seurat@meta.data <- meta

DimPlot(spl_seurat, group.by = "cluster_annotation", label = TRUE, repel = TRUE) + NoLegend()
```

```{r}
# ploteamos los porcentajes
plot_integrated_clusters_annotated(spl_seurat)
```

### specific H2AX

```{r}
# if i want to check specific genes in cluster
# volcano plot 
VlnPlot(
  spl_seurat,
  features = "H2ax",
  group.by = "cluster_annotation",   # usa la anotación en lugar del número
  split.by = "orig.ident"            # WT vs KO
) + 
  ggtitle("H2ax expression by annotated clusters (WT vs KO)")
```



## Manual annotation

```{r}
# para buscar marcadores de cada cluster:
DefaultAssay(spl_seurat) <- "RNA"
spl_seurat <- JoinLayers(spl_seurat)
# findallmarkers
markers_spl <- FindAllMarkers(spl_seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
head(markers_spl)
# write.csv(markers_spl, "markers_spleen.csv")
```

Ahora hacemos un string de todos:

```{r}
# Paso 1: Define las anotaciones para los 16 clústeres
annotations <- c(
  "B cells",                    # cluster 0
  "T cells",                    # cluster 1
  "T cell activation",          # cluster 2
  "Neutrophils",                # cluster 3
  "T cells",                    # cluster 4
  "Vesicle-mediated transport", # cluster 5
  "Antigen presentation",       # cluster 6
  "T cell activation",          # cluster 7
  "ER-stress",                  # cluster 8
  "Chromosome seggregation",    # cluster 9
  "B cell activation",          # cluster 10
  "Neutrophil activation",      # cluster 11
  "NK cells",                   # cluster 12
  "T cells/ribosomes",          # cluster 13
  "Myeloid cell differentiation", # cluster 14
  "Leukocyte/granulocyte"       # cluster 15
)

# Asegurar que las anotaciones estén bien nombradas
names(annotations) <- as.character(0:15)

# Paso 2: Obtener el vector de clúster por celda
cluster_ids <- spl_seurat$seurat_clusters
names(cluster_ids) <- colnames(spl_seurat)

# Paso 3: Mapear anotaciones por celda
cluster_annotations <- annotations[as.character(cluster_ids)]

# Paso 4: Añadir al metadata
spl_seurat@meta.data$cluster_annotation_string <- cluster_annotations

# Visualizar
DimPlot(spl_seurat, group.by = "cluster_annotation_string", label = TRUE, repel = TRUE) + NoLegend()
```

## Markers:

2 kind of plots:

```{r}
# para chequear marcadores:
FeaturePlot(spl_seurat, features = c("Adgre1", "Cd68", "Hba-a1", "Ly6g"), label = TRUE)
```

```{r}
VlnPlot(spl_seurat, features = c("Adgre1", "Ly6g", "Klf1"), group.by = "cluster_annotation")
```



# Bone marrow

## Integration, normalization

```{r}
# integration
# creat a list
bm_list <- list()
bm_list[["bm_WT"]] <- bm_WT_obj
bm_list[["bm_KO"]] <- bm_KO_obj

# normalize/find HVG
for (i in 1:length(bm_list)) {
  bm_list[[i]] <- NormalizeData(bm_list[[i]], verbose = F)
  bm_list[[i]] <- FindVariableFeatures(bm_list[[i]], selection.method = "vst", nfeatures = 2000, verbose = F)
}
# to find integration anchors and actually perform integration
# it takes time...
bm_anchors <- FindIntegrationAnchors(object.list = bm_list, dims = 1:30)
bm_seurat <- IntegrateData(anchorset = bm_anchors, dims = 1:30)
# to save ram
rm(bm_list)
rm(bm_anchors)

# before integration
DefaultAssay(bm_seurat) <- "RNA"
bm_seurat <- NormalizeData(bm_seurat, verbose = F)
bm_seurat <- FindVariableFeatures(bm_seurat, selection.method = "vst", nfeatures = 2000, verbose = F)
bm_seurat <- ScaleData(bm_seurat, verbose = F)
bm_seurat <- RunPCA(bm_seurat, npcs = 30, verbose = F)
bm_seurat <- RunUMAP(bm_seurat, reduction = "pca", dims = 1:30, verbose = F)

DimPlot(bm_seurat,reduction = "umap") +
  plot_annotation(title = "Bone marrow, before integration")
```

After integration:

```{r}
# the integrated assay (it’s already normalized and HVGs are selected)
DefaultAssay(bm_seurat) <- "integrated"
bm_seurat <- ScaleData(bm_seurat, verbose = F)
bm_seurat <- RunPCA(bm_seurat, npcs = 10, verbose = F)
bm_seurat <- RunUMAP(bm_seurat, reduction = "pca", dims = 1:10, verbose = F)

DimPlot(bm_seurat, reduction = "umap") + 
  plot_annotation(title = "Bone marrow, after integration (Seurat 3)")
```

## Clusterization

```{r}
# let's cluster
bm_seurat <- FindNeighbors(bm_seurat, dims = 1:10, k.param = 10, verbose = F)
# less resolution, less clusters
bm_seurat <- FindClusters(bm_seurat, verbose = F, resolution = 0.1)
DimPlot(bm_seurat,label = T) + NoLegend()
```

```{r}
plot_integrated_clusters(bm_seurat)
```

## Annotation

### Immuno dataset

```{r}
### annotation

# Extraer las matrices de cada muestra
expr_WT_bm <- GetAssayData(bm_seurat, assay = "RNA", layer = "data.bm_WT")
expr_KO_bm <- GetAssayData(bm_seurat, assay = "RNA", layer = "data.bm_KO")

# Combina columnas (células) de ambos
expr_bm <- cbind(expr_WT_bm, expr_KO_bm)


bm_sce <- SingleCellExperiment(
  assays = list(logcounts = expr_bm),
  colData = bm_seurat@meta.data
)

# mouse inmune reference
ref <- celldex::ImmGenData()

# Corre SingleR para anotar
pred_bm <- SingleR(test = bm_sce, ref = ref, labels = ref$label.fine)

# Añade la anotación a tu objeto Seurat
bm_seurat$SingleR_labels <- pred_bm$labels
# limpiamos
bm_seurat$SingleR_clean <- gsub("\\s*\\([^\\)]+\\)", "", bm_seurat$SingleR_labels)

# Visualiza con nombres más cortos
DimPlot(bm_seurat, group.by = "SingleR_clean", label = TRUE, repel = TRUE) + NoLegend()
```

To annotate clusters:

```{r}
# para anotar los clusteres

# 1. Extraer metadata
meta <- bm_seurat@meta.data

# 2. Crear tabla con la anotación más frecuente por clúster
annot_df <- meta %>%
  mutate(cluster = as.character(seurat_clusters)) %>%
  group_by(cluster) %>%
  count(SingleR_clean) %>%
  slice_max(n, n = 1, with_ties = FALSE) %>%
  ungroup()

# 3. Crear un vector de anotación por clúster
annot_vec <- setNames(annot_df$SingleR_clean, annot_df$cluster)

# 4. Asignar anotaciones a cada célula usando su clúster
meta$cluster_annotation <- annot_vec[as.character(meta$seurat_clusters)]

# 5. Asegurarse de que se guarda en el objeto Seurat
bm_seurat@meta.data <- meta

DimPlot(bm_seurat, group.by = "cluster_annotation", label = TRUE, repel = TRUE) + NoLegend()
```

```{r}
# ploteamos los porcentajes
plot_integrated_clusters_annotated(bm_seurat)
```

gH2AX (well in mice, is h2ax)

```{r}
# if i want to check specific genes in cluster
# volcano plot 
VlnPlot(
  bm_seurat,
  features = "H2ax",
  group.by = "cluster_annotation",   # usa la anotación en lugar del número
  split.by = "orig.ident"            # WT vs KO
) + 
  ggtitle("H2ax expression by annotated clusters (WT vs KO)")
```


### Mousernaseq dataset

```{r}
# mouse inmune reference
ref <- celldex::MouseRNAseqData()

# Corre SingleR para anotar
pred_bm <- SingleR(test = bm_sce, ref = ref, labels = ref$label.fine)

# Añade la anotación a tu objeto Seurat
bm_seurat$SingleR_labels <- pred_bm$labels
# limpiamos
bm_seurat$SingleR_clean <- gsub("\\s*\\([^\\)]+\\)", "", bm_seurat$SingleR_labels)

# Visualiza con nombres más cortos
DimPlot(bm_seurat, group.by = "SingleR_clean", label = TRUE, repel = TRUE) + NoLegend()


# para anotar los clusteres

# 1. Extraer metadata
meta <- bm_seurat@meta.data

# 2. Crear tabla con la anotación más frecuente por clúster
annot_df <- meta %>%
  mutate(cluster = as.character(seurat_clusters)) %>%
  group_by(cluster) %>%
  count(SingleR_clean) %>%
  slice_max(n, n = 1, with_ties = FALSE) %>%
  ungroup()

# 3. Crear un vector de anotación por clúster
annot_vec <- setNames(annot_df$SingleR_clean, annot_df$cluster)

# 4. Asignar anotaciones a cada célula usando su clúster
meta$cluster_annotation <- annot_vec[as.character(meta$seurat_clusters)]

# 5. Asegurarse de que se guarda en el objeto Seurat
bm_seurat@meta.data <- meta

DimPlot(bm_seurat, group.by = "cluster_annotation", label = TRUE, repel = TRUE) + NoLegend()

# ploteamos los porcentajes
plot_integrated_clusters_annotated(bm_seurat)
```


## Manual annotation

```{r}
# para buscar marcadores de cada cluster:

DefaultAssay(bm_seurat) <- "RNA"
bm_seurat <- JoinLayers(bm_seurat)
# findallmarkers
markers_bm <- FindAllMarkers(bm_seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
head(markers_bm)
#write.csv(markers_bm, "OUTPUT/markers_bonemarrow.csv")
```


```{r}
# Paso 1: Define las anotaciones para los 13 clústeres
annotations <- c(
  "Ribosomal-B cells",          # cluster 0
  "T cells",                    # cluster 1
  "T cell activation",          # cluster 2
  "Neutrophils/Macrophages",    # cluster 3
  "T cells",                    # cluster 4
  "Vesicle-mediated transport", # cluster 5
  "Antigen presentation/Macrophages",# cluster 6
  "T cell activation",          # cluster 7
  "ER-stress/Pancreas",         # cluster 8
  "Chromosome seggregation",    # cluster 9
  "B cell activation",          # cluster 10
  "Neutrophil activation",      # cluster 11
  "NK cells",                   # cluster 12
  "T cell",                     # cluster 13
  "Myeloid cell differentiation",      # cluster 14
  "Granulocyte"                 # cluster 15
)

# Asegurar que las anotaciones estén bien nombradas
names(annotations) <- as.character(0:15)

# Paso 2: Obtener el vector de clúster por celda
cluster_ids <- bm_seurat$seurat_clusters
names(cluster_ids) <- colnames(bm_seurat)

# Paso 3: Mapear anotaciones por celda
cluster_annotations <- annotations[as.character(cluster_ids)]

# Paso 4: Añadir al metadata
bm_seurat@meta.data$cluster_annotation_string <- cluster_annotations

# Visualizar
DimPlot(bm_seurat, group.by = "cluster_annotation_string", label = TRUE, repel = TRUE) + NoLegend()
```



# R session

```{r}
sessionInfo()
```

